gxtest/copyfilter.cpp

```diff
@@ -0,0 +1,239 @@
+  if (copy_filter_sum > 63)
+    w3 = std::min<u8>(copy_filter_sum - 63, 63);
+  u8 w5 = 0;
+  if (copy_filter_sum > 63*2)
+    w5 = std::min<u8>(copy_filter_sum - 63 * 2, 63);
+
+  u32 copy_filter_reg_0 = u32(w3) << 12 | u32(w4) << 18;
+  u32 copy_filter_reg_1 = u32(w5);
+
+  CGX_LOAD_BP_REG(BPMEM_COPYFILTER0 << 24 | copy_filter_reg_0);
+  CGX_LOAD_BP_REG(BPMEM_COPYFILTER1 << 24 | copy_filter_reg_1);
+}
+
+GXTest::Vec4<u8> GetEfbColor(u8 x, u32 pixel_fmt)
+{
+  // const u8 sixbit = static_cast<u8>(((x & 0xfc) * 255) / 0xfc);
+  // const u8 fivebit = static_cast<u8>(((x & 0xf8) * 255) / 0xf8);
+  const u8 sixbit = static_cast<u8>((x & 0xfc) | ((x & 0xc0) >> 6));
+  const u8 fivebit = static_cast<u8>((x & 0xf8) | ((x & 0xe0) >> 5));
+  switch (pixel_fmt)
+  {
+  case PIXELFMT_RGB8_Z24:
+  case PIXELFMT_Y8:
+  case PIXELFMT_YUV420:
+  default:
+    return {x, x, x, 255};
+  case PIXELFMT_RGBA6_Z24:
+    return {sixbit, sixbit, sixbit, sixbit};
+  case PIXELFMT_RGB565_Z16:
+    // Does not work
+    return {fivebit, sixbit, fivebit, 255};
+  case PIXELFMT_U8:
+  case PIXELFMT_V8:
+    // Does not work
+    return {fivebit, fivebit, fivebit, 255};
+  case PIXELFMT_Z24:
+    // Does not work
+    return {x, 0, 0, 255};
+  }
+}
+
+u8 Predict(u8 value, u8 copy_filter_sum, u32 gamma)
+{
+  // Apply copy filter
+  u32 prediction_i = static_cast<u32>(value) * static_cast<u32>(copy_filter_sum);
+  prediction_i >>= 6;  // Divide by 64
+  // The clamping seems to happen in the range[0, 511]; if the value is outside
+  // an overflow will still occur.  This happens if copy_filter_sum >= 128.
+  prediction_i &= 0x1ffu;
+  prediction_i = std::min(prediction_i, 0xffu);
+  // Apply gamma
+  if (gamma != GAMMA_1_0)
+  {
+    // Convert from [0-255] to [0-1]
+    float prediction_f = static_cast<float>(prediction_i) / 255.f;
+    switch (gamma)
+    {
+    case GAMMA_1_7:
+      prediction_f = std::pow(prediction_f, 1 / 1.7f);
+      break;
+    case GAMMA_2_2:
+    case GAMMA_INVALID_2_2:
+    default:
+      prediction_f = std::pow(prediction_f, 1 / 2.2f);
+      break;
+    }
+    // Due to how exponentials work, std::pow will always map from [0, 1] to [0, 1],
+    // so no overflow can occur here.  (pow is continuous, 0^x is 0 for x > 0,
+    // and 1^x is 1, so y in [0, 1] has y^x in [0, 1])
+    // Convert back from [0, 1] to [0, 255]
+    prediction_i = static_cast<u32>(std::round(prediction_f * 255.f));
+  }
+  return static_cast<u8>(prediction_i);
+}
+
+GXTest::Vec4<u8> Predict(GXTest::Vec4<u8> efb_color, u8 copy_filter_sum, u32 gamma, bool intensity)
+{
+  const u8 r = Predict(efb_color.r, copy_filter_sum, gamma);
+  const u8 g = Predict(efb_color.g, copy_filter_sum, gamma);
+  const u8 b = Predict(efb_color.b, copy_filter_sum, gamma);
+  const u8 a = efb_color.a;  // Copy filter doesn't apply to alpha
+  if (intensity)
+  {
+    // BT.601 conversion
+    const u8 y = static_cast<u8>(std::round( 0.2578125f * r + 0.50390625f * g + 0.09765625f * b + 16));
```

Comment:
Can those exact coefficients be derived purely from gamma==1.0, or did you also use information from other gamma factors? If the latter, how did you decide whether to change the yuv coefficients or the prediction formula?

Reply:
They are derived purely from gamma == 1.0 and the copy filter set to 64 (see the intensity test).  After plugging in the ones from that test, they worked correctly for other gamma and copy filter factors.

Reply:
Great, thanks for clarifying :)